examples/llama2/fsdp_finetune.py

```diff
@@ -0,0 +1,250 @@
+    input_ids = examples_tokenized['input_ids']
+    labels = copy.deepcopy(input_ids)
+    for label, source_len in zip(labels, sources_tokenized['input_ids_lens']):
+        label[:source_len] = IGNORE_INDEX
+    return dict(input_ids=input_ids, labels=labels)
+
+
+class SupervisedDataset(Dataset):
+    """""Dataset for supervised fine-tuning.""."""
+
+    def __init__(self, data_path: str,
+                 tokenizer: transformers.PreTrainedTokenizer):
+        super().__init__()
+        print_log('Loading data...', level=logging.WARNING)
+        list_data_dict = load(data_path)
+
+        print_log('Formatting inputs...', level=logging.WARNING)
+        prompt_input, prompt_no_input = \
+            PROMPT_DICT['prompt_input'], PROMPT_DICT['prompt_no_input']
+        sources = [
+            prompt_input.format_map(example) if example.get('input', '') != ''
+            else prompt_no_input.format_map(example)
+            for example in list_data_dict
+        ]
+        targets = [
+            f'{example["output"]}{tokenizer.eos_token}'
+            for example in list_data_dict
+        ]
+
+        print_log(
+            'Tokenizing inputs... This may take some time...',
+            level=logging.WARNING)
+        data_dict = preprocess(sources, targets, tokenizer)
+
+        self.input_ids = data_dict['input_ids']
+        self.labels = data_dict['labels']
+
+    def __len__(self):
+        return len(self.input_ids)
+
+    def __getitem__(self, i) -> Dict[str, torch.Tensor]:
+        return dict(input_ids=self.input_ids[i], labels=self.labels[i])
+
+
+@dataclass
+class DataCollatorForSupervisedDataset:
+    """Collate examples for supervised fine-tuning."""
+
+    tokenizer: transformers.PreTrainedTokenizer
+
+    def __call__(self, instances: Sequence[Dict]) -> Dict[str, torch.Tensor]:
+        input_ids, labels = tuple([instance[key] for instance in instances]
+                                  for key in ('input_ids', 'labels'))
+        input_ids = torch.nn.utils.rnn.pad_sequence(
+            input_ids,
+            batch_first=True,
+            padding_value=self.tokenizer.pad_token_id)
+        labels = torch.nn.utils.rnn.pad_sequence(
+            labels, batch_first=True, padding_value=IGNORE_INDEX)
+        attention_mask = input_ids.ne(  # type: ignore
+            self.tokenizer.pad_token_id)
+        return dict(
+            input_ids=input_ids,
+            labels=labels,
+            attention_mask=attention_mask,
+        )
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Distributed Training')
+    parser.add_argument('data_root', type=str)
+    parser.add_argument('checkpoint', type=str)
+    parser.add_argument('--output-dir', type=str, default='work_dirs')
+    parser.add_argument('--max-epoch', type=int, default=2)
+    parser.add_argument('--batch-size', type=int, default=2)
+    args = parser.parse_args()
+    return args
+
+
+def train():
```

Comment:
I suggest adding additional comments and documentation references to explain each step and core modules like strategy. I think this demo will be the entry point for some users interested in Llama v2 but no familiar with the mmengine strategy.